saq/queue/postgres.py

```diff
@@ -0,0 +1,521 @@
+
+        if status == Status.COMPLETE:
+            self.complete += 1
+        elif status == Status.FAILED:
+            self.failed += 1
+        elif status == Status.ABORTED:
+            self.aborted += 1
+
+        await self.notify(job)
+        logger.info("Finished %s", job.info(logger.isEnabledFor(logging.DEBUG)))
+
+    async def dequeue(self, timeout: float = 0) -> Job | None:
+        try:
+            return await asyncio.wait_for(self.wait_for_job(), timeout or None)
+        except asyncio.exceptions.TimeoutError:
+            return None
+
+    async def enqueue(self, job_or_func: str | Job, **kwargs: t.Any) -> Job | None:
+        job = self._create_job_for_enqueue(job_or_func, **kwargs)
+
+        await self._before_enqueue(job)
+
+        async with self.pool.connection() as conn, conn.cursor() as cursor:
+            await cursor.execute(
+                f"""
+                INSERT INTO {self.jobs_table} (key, job, queue, status, scheduled)
+                VALUES (%(key)s, %(job)s, %(queue)s, %(status)s, %(scheduled)s)
+                ON CONFLICT (key) DO UPDATE
+                SET job = %(job)s, queue = %(queue)s, status = %(status)s, scheduled = %(scheduled)s
+                WHERE {self.jobs_table}.status NOT IN ('queued', 'aborting')
+                RETURNING job
+                """,
+                {
+                    "key": job.key,
+                    "job": self.serialize(job),
+                    "queue": self.name,
+                    "status": job.status,
+                    "scheduled": job.scheduled or seconds(now()),
+                },
+            )
+
+            if not await cursor.fetchone():
+                return None
+
+        await self._notify("saq:enqueue", job.key)
+        logger.info("Enqueuing %s", job.info(logger.isEnabledFor(logging.DEBUG)))
+        return job
+
+    async def get_abort_errors(self, jobs: Iterable[Job]) -> list[bytes | None]:
+        async with self.pool.connection() as conn, conn.cursor() as cursor:
+            await cursor.execute(
+                f"""
+                SELECT key, job->'error'
+                FROM {self.jobs_table}
+                WHERE key = ANY(%(keys)s)
+                """,
+                {"keys": [job.key for job in jobs]},
+            )
+            results: dict[str, str | None] = dict(await cursor.fetchmany())
+        errors = []
+        for job in jobs:
+            error = results.get(job.key)
+            errors.append(error.encode("utf-8") if error else None)
+        return errors
+
+    async def finish_abort(self, job: Job) -> None:
+        await job.finish(Status.ABORTED, error=job.error)
+
+    async def dequeue_timer(self, poll_interval: int) -> None:
+        """Wakes up a single dequeue task every `poll_interval` seconds."""
+        while True:
+            async with self.cond:
+                self.cond.notify(1)
+            await asyncio.sleep(poll_interval)
+
+    async def wait_for_job(self) -> Job:
```

Comment:
i feel like this isn't necessary, it's pretty much the same as dequeue, can we move it inline into dequeu?

Reply:
Do you mean make this an inner function in dequeue()? This was added so dequeue() can use asyncio.wait_for() for timing out, but this can be nested inside dequeue().

Reply:
yea lambda inside